### ISSUE TITLE
SURF_CUDA produces different number of keypoints on the same image

### ISSUE THREAD

**Issue by drtimes** [opencv/opencv_contrib#584]
For some images at each run SURF_CUDA() produces different number of detected keypoints on the same image, no matter if it is the same instance or another instance. I'm using OpenCV 3.1.0, CUDA 7.5

![seaman_surf-cuda.jpg](https://cloud.githubusercontent.com/assets/13232153/13806850/c44caa56-eb66-11e5-9023-3e018b0651de.jpg)

Sample image: [seaman.jpg](https://cloud.githubusercontent.com/assets/13232153/13806851/c44ce9ee-eb66-11e5-906b-ef91a08e44fd.jpg)

```
#include <iostream>
#include "opencv2/core/core.hpp"
#include "opencv2/features2d/features2d.hpp"
#include "opencv2/highgui/highgui.hpp"
#include "opencv2/cudafeatures2d.hpp"
#include "opencv2/xfeatures2d/cuda.hpp"
#include "opencv2/opencv_modules.hpp"

using namespace std;
using namespace cv;
using namespace cv::cuda;

int main(int argc, char *argv[])
{
    setDevice(0);

    GpuMat img1, img2;
    img1.upload(imread("seaman.jpg", IMREAD_GRAYSCALE));
    img2.upload(imread("seaman.jpg", IMREAD_GRAYSCALE));

    SURF_CUDA surf;
    GpuMat keypoints1GPU, keypoints2GPU;
    GpuMat descriptors1GPU, descriptors2GPU;

    surf(img1, GpuMat(), keypoints1GPU, descriptors1GPU);
    surf(img2, GpuMat(), keypoints2GPU, descriptors2GPU);

    cout << "FOUND " << keypoints1GPU.cols << " keypoints on first image" << endl;
    cout << "FOUND " << keypoints2GPU.cols << " keypoints on second image" << endl;

    Ptr<DescriptorMatcher> matcher = DescriptorMatcher::createBFMatcher(surf.defaultNorm());
    vector<DMatch> matches;
    matcher->match(descriptors1GPU, descriptors2GPU, matches);

    vector<KeyPoint> keypoints1, keypoints2;
    vector<float> descriptors1, descriptors2;
    surf.downloadKeypoints(keypoints1GPU, keypoints1);
    surf.downloadKeypoints(keypoints2GPU, keypoints2);
    surf.downloadDescriptors(descriptors1GPU, descriptors1);
    surf.downloadDescriptors(descriptors2GPU, descriptors2);

    // draw the results
    Mat img_matches;
    drawMatches(Mat(img1), keypoints1, Mat(img2), keypoints2, matches, img_matches);
    imwrite("seaman_result.jpg", img_matches);

    namedWindow("matches", 0);
    imshow("matches", img_matches);
    waitKey(0);

    resetDevice();

    return 0;
}
```

Another example: [graf3.png](https://cloud.githubusercontent.com/assets/13232153/13806886/fbdf65c6-eb66-11e5-9bf6-d69487d3d298.png) [graf3_surf-cuda.jpg](https://cloud.githubusercontent.com/assets/13232153/13806849/c44c1140-eb66-11e5-9720-2f1f49e5774d.jpg)

**Comment by Bleach665** [comment]
@alalek, @mshabunin. Also SURF in Opencv 3.1 and in 2.4.13 produces different number of keypoints. Is this a bug? Create issues?

**Comment by bmstoyon** [comment]
I notice the name issue recently. I ran a test compare results while varying the image size and discover that every 3 out of 4 multiples of 16 is not repeatable.

For example the following size are all non repeatable.
[wdith x height]
[304 x 4000]
[336 x 4000]
[352 x 4000]
[368 x 4000]
[400 x 4000]
[416 x 4000]
[432 x 4000]
[464 x 4000]
[480 x 4000]
[496 x 4000]
[528 x 4000]
[544 x 4000]
[560 x 4000]

I believe the bug is actually in cv::cuda::integral. Which may or may not have been fixed already.

**Comment by bmstoyon** [comment]
I proved that there is an issue with cuda::integral. And issue has been added to the core repo. https://github.com/opencv/opencv/issues/7378

**Comment by willard-yuan** [comment]
I also faced the problem~

**Comment by catree** [comment]
With OpenCV  fc9e031 and OpenCV contrib at 0327b79:

![result_seaman_master](https://user-images.githubusercontent.com/8229425/33532681-fb428962-d89b-11e7-9c2f-8cf421c64257.png)

> FOUND 765 keypoints on first image
> FOUND 765 keypoints on second image

![result_graf3_master](https://user-images.githubusercontent.com/8229425/33532684-042b03d8-d89c-11e7-904b-d0b038211e6d.png)

> FOUND 4696 keypoints on first image
> FOUND 4696 keypoints on second image

With fix from opencv/opencv#7378:

![result_seaman_fixed](https://user-images.githubusercontent.com/8229425/33532704-36f391a4-d89c-11e7-8728-94038b96c31a.png)

> FOUND 513 keypoints on first image
> FOUND 513 keypoints on second image

![result_graf3_fixed](https://user-images.githubusercontent.com/8229425/33532713-491991c6-d89c-11e7-9571-3b93dce8e316.png)

> FOUND 4628 keypoints on first image
> FOUND 4628 keypoints on second image

**Comment by savuor** [comment]
Was resolved by [#10218@opencv](https://github.com/opencv/opencv/pull/10218)
